File: webapp/src/main/scala/org/linnando/mixemulator/webapp/LineAccessFileInputOpsImpl.scala
package org.linnando.mixemulator.webapp

import io.scalajs.nodejs.fs.{FileInputOptions, Fs}
import org.linnando.mixemulator.vm.exceptions.EndOfFileException
import org.linnando.mixemulator.vm.io.file.LineAccessFileInputOps

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

class LineAccessFileInputOpsImpl extends LineAccessFileInputOps {
  override def readLine(filename: String, position: Long): Future[Array[Char]] =
    Fs.readFileFuture(filename, new FileInputOptions(encoding = "utf8")).map(data => {
      val lines = data.asInstanceOf[String].split("\n")
      if (position < lines.length) lines(position.toInt).toCharArray
      else throw new EndOfFileException
    })

  override def initialise(filename: String): Future[Unit] =
    for {
      exists <- Fs.existsFuture(filename)
      _ <- if (exists) Future {} else save(filename, "")
    } yield ()

  override def save(filename: String, data: String): Future[Unit] =
    Fs.writeFileFuture(s"/$filename", data)

  override def getData(filename: String): Future[String] =
    Fs.readFileFuture(s"/$filename", new FileInputOptions(encoding = "utf8")).map(_.asInstanceOf[String])
}

File: webapp/src/app/line-access-file-input-ops-impl.ts
import { EndOfFileException, LineAccessFileInputOps } from '@mixemulator/lib';

export class LineAccessFileInputOpsImpl extends LineAccessFileInputOps {
  constructor(private storage: Storage) {
    super();
  }

  override readLine(filename: string, position: number): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const data = this.storage.getItem(filename);
        if (data == null) {
          return reject(new Error(`File ${filename} not found.`));
        }
        const lines = data.split('\n');
        if (position < lines.length) {
          return resolve(lines[position]);
        } else {
          return reject(new EndOfFileException());
        }
      } catch (e) {
        return reject(e);
      }
    });
  }

  override initialise(filename: string): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        if (this.storage.getItem(filename) == null) {
          this.storage.setItem(filename, '');
        }
        return resolve();
      } catch (e) {
        return reject(e);
      }
    });
  }

  override save(filename: string, data: string): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.storage.setItem(filename, data);
        return resolve();
      } catch (e) {
        return reject(e);
      }
    });
  }

  override getData(filename: string): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const data = this.storage.getItem(filename);
        if (data == null) {
          return reject(new Error(`File ${filename} not found.`));
        }
        return resolve(data);
      } catch (e) {
        return reject(e);
      }
    });
  }
}
